libnavui-util/src/main/java/com/mapbox/navigation/ui/utils/resource/DefaultResourceLoader.kt

```diff
@@ -0,0 +1,88 @@
+package com.mapbox.navigation.ui.utils.resource
+
+import com.mapbox.bindgen.Expected
+import com.mapbox.common.Cancelable
+import com.mapbox.common.ResourceDescription
+import com.mapbox.common.ResourceLoadError
+import com.mapbox.common.ResourceLoadOptions
+import com.mapbox.common.ResourceLoadProgress
+import com.mapbox.common.ResourceLoadProgressCallback
+import com.mapbox.common.ResourceLoadResult
+import com.mapbox.common.ResourceLoadResultCallback
+import com.mapbox.common.TileDataDomain
+import com.mapbox.common.TileStore
+import java.util.Queue
+import java.util.concurrent.ConcurrentHashMap
+import java.util.concurrent.ConcurrentLinkedQueue
+import java.util.concurrent.atomic.AtomicLong
+
+/**
+ * Default [ResourceLoader] implementation.
+ *
+ * Uses [TileStore] to load resources.
+ * Keeps track of all [Cancelable] instances returned by the [TileStore.loadResource] and
+ * assigns each a unique <i>id</i> that can be passed to [ResourceLoader.cancel] to
+ * abort load operation.
+ */
+internal class DefaultResourceLoader(
+    private val tileStore: TileStore
+) : ResourceLoader() {
+
+    private val nextRequestId = AtomicLong(0L)
+    private val cancelableMap = ConcurrentHashMap<Long, Cancelable>()
+    private val observers: Queue<ResourceLoadObserver> = ConcurrentLinkedQueue()
+
+    override fun load(request: ResourceLoadRequest, callback: ResourceLoadCallback): Long {
+        val requestId = nextRequestId.incrementAndGet()
+        val callbackAdapter = CallbackAdapter(request, callback, *observers.toTypedArray())
```

Comment:
Passing an immutable collection to the adapter would prevent one from listening to the progress of a request if the observer was registered after the request was dispatched, which doesn't look correct.

An observer should be notified with all available data points regardless of when it was registered.

Reply:
Good catch.